src/threaddeadlock.cpp

```diff
@@ -0,0 +1,718 @@
+        // we dont need to check empty here, other thread has at least 1 lock otherwise we wouldnt be checking it
+        if (other_iter->second.size() == 1)
+        {
+            // it does not have any locks aside from known exclusive, no deadlock possible
+            // we can just wait until that exclusive lock is released
+            return false;
+        }
+        // if the other thread has 1+ other locks aside from the known exclusive, check them for matches with our own
+        // locks
+        for (auto &lock : other_iter->second)
+        {
+            // if they have a lock that is on a lock that we have exclusive ownership for
+            if (HasAnyOwners(lock.first))
+            {
+                // and their lock is waiting...
+                if (lock.second.GetWaiting() == true)
+                {
+                    deadlocks.push_back(lock);
+                    threads.emplace(other_iter->first);
+                    if (other_iter->first == tid && lock.first == c)
+                    {
+                        // we are back where we started and there is a deadlock
+                        return true;
+                    }
+                    if (ReadRecursiveCheck(tid, c, other_iter->first, lock.first, false, deadlocks, threads))
+                    {
+                        return true;
+                    }
+                }
+                // no deadlock, other lock is not waiting, we simply have to wait until they release that lock
+            }
+        }
+    }
+    return false;
+}
+
+static bool WriteRecursiveCheck(const uint64_t &tid,
+    const void *c,
+    uint64_t lastTid,
+    void *lastLock,
+    bool firstRun,
+    LockStack &deadlocks,
+    std::set<uint64_t> &threads)
+{
+    if (!firstRun && c == lastLock && tid == lastTid)
+    {
+        // we are back where we started, infinite loop means there is a deadlock
+        return true;
+    }
+    // first check if we currently have any exclusive ownerships
+    size_t selfOtherLockCount = 0;
+    auto self_iter = lockdata.locksheldbythread.find(lastTid);
+    if (self_iter != lockdata.locksheldbythread.end() && self_iter->second.empty() == false)
+    {
+        selfOtherLockCount = self_iter->second.size();
+    }
+    // we cant deadlock if we dont own any other mutexs
+    if (selfOtherLockCount == 0)
+    {
+        return false;
+    }
+    // at this point we have at least 1 lock for a mutex somewhere
+
+    // check if a thread has an ownership of c
+    auto writeiter = lockdata.writelocksheld.find(lastLock);
+    auto readiter = lockdata.readlocksheld.find(lastLock);
+
+    // NOTE: be careful when adjusting these booleans, the order of the checks is important
+    bool readIsEnd = ((readiter == lockdata.readlocksheld.end()) || readiter->second.empty());
+    bool writeIsEnd = ((writeiter == lockdata.writelocksheld.end()) || writeiter->second.empty());
+    if (writeIsEnd && readIsEnd)
```

Comment:
Nit: `readIsEnd` was not intuitive to me, how about `isReadLocked` (and reverse the boolean value)

Reply:
agreed, fixed